apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/SizedBufferingConsumer.cs

```diff
@@ -61,42 +63,97 @@ private SizedBufferingConsumer(IConsumer<T> consumer, ISizer<T> sizer, int buffe
         /// <param name="consumer">The consumer to flush to, cannot be null.</param>
         /// <param name="bufferSize">The buffer size in bytes.</param>
         public static SizedBufferingConsumer<T> Create(IConsumer<T> consumer, ISizer<T> sizer, int bufferSize)
+            => new SizedBufferingConsumer<T>(consumer, sizer, bufferSize);
+
+        /// <inheritdoc />
+        public void Receive(IEnumerable<T> items)
         {
-            return new SizedBufferingConsumer<T>(consumer, sizer, bufferSize);
+            GaxPreconditions.CheckNotNull(items, nameof(items));
+            _semaphore.Wait();
+            foreach (T item in items)
+            {
+                _size += _sizer.GetSize(item);
+                _items.Add(item);
+                if (_size >= _bufferSize)
+                {
+                    Flush(waitForSemaphore: false);
+                }
+            }
+            _semaphore.Release();
         }
 
         /// <inheritdoc />
-        public void Receive(IEnumerable<T> items)
+        public async Task ReceiveAsync(
+            IEnumerable<T> items, CancellationToken cancellationToken = default(CancellationToken))
         {
             GaxPreconditions.CheckNotNull(items, nameof(items));
-            lock (_mutex)
+            try
             {
+                await _semaphore.WaitAsync(cancellationToken);
                 foreach (T item in items)
                 {
                     _size += _sizer.GetSize(item);
                     _items.Add(item);
                     if (_size >= _bufferSize)
                     {
-                        Flush();
+                        await FlushAsync(cancellationToken, waitForSemaphore: false);
                     }
                 }
             }
+            finally
+            {
+                _semaphore.Release();
+            }
         }
 
         /// <inheritdoc />
-        public void Flush()
+        public void Flush() => Flush(waitForSemaphore: true);
+
+        /// <inheritdoc />
+        public async Task FlushAsync(CancellationToken cancellationToken = default(CancellationToken)) => 
+            await FlushAsync(cancellationToken, waitForSemaphore: true);
+
+        /// <summary>
+        /// Flush the buffer.
+        /// </summary>
+        /// <param name="waitForSemaphore">If true the flush will wait for the semaphore.</param>
+        private void Flush(bool waitForSemaphore)
```

Comment:
I would refactor this into two methods: a Flush method that takes the semaphore and then calls the other method,  FlushWithSemaphoreHeld which can be called directly elsewhere.